### PR TITLE
docs: add usage and API design in README.md

### LINKED ISSUE #4: docs: moving forward with this library
#### Reason/Context
Fortunately, the project is initialized and we can discuss the API of this library here.

#### Description
Here is a draft of the API as well as the proposed Report structure.
After finalizing the API and report structure, I will open a pull request with these changes in the README.md file.

### Node.js

```typescript
import { Optimizer } from '@asyncapi/optimizer';
import { parse } from '@asyncapi/parser';
let asyncApiDocument = parse(`
asyncapi: 2.0.0
info:
  title: Streetlights API
  version: '1.0.0'

channels:

  smartylighting/event/{streetlightId}/lighting/measured:
    parameters:
      #this parameter is duplicated. it can be moved to components and ref-ed from here.
      streetlightId:
        schema:
          type: string
    subscribe:
      operationId: receiveLightMeasurement
      traits:
        - bindings:
            kafka:
              clientId: my-app-id
      message:
        name: lightMeasured
        title: Light measured
        contentType: application/json
        traits:
          - headers:
              type: object
              properties:
                my-app-header:
                  type: integer
                  minimum: 0
                  maximum: 100
        payload:
          type: object
          properties:
            lumens:
              type: integer
              minimum: 0
            #full form is used, we can ref it to: #/components/schemas/sentAt
            sentAt:
              type: string
              format: date-time

  smartylighting/action/{streetlightId}/turn/on:
    parameters:
      streetlightId:
        schema:
          type: string
    publish:
      operationId: turnOn
      traits:
        - bindings:
            kafka:
              clientId: my-app-id
      message:
        name: turnOnOff
        title: Turn on/off
        traits:
          - headers:
              type: object
              properties:
                my-app-header:
                  type: integer
                  minimum: 0
                  maximum: 100
        payload:
          type: object
          properties:
            sentAt:
              $ref: "#/components/schemas/sentAt"

components:
  messages:
    #libarary should be able to find and delete this message because it is not used anywhere.
    unusedMessage:
      name: unusedMessage
      title: This message is not used in any channel.
      
  schemas:
    #this schema is ref-ed in one channel and used full form in another. the library should be able to identify and ref the second channel as well.
    sentAt:
      type: string
      format: date-time`
);
let optimizer = new Optimizer(asyncApiDocument);
let report: OptimizerReport = await optimizer.getReport();
/*
the `report` value will be:
{
  ReuseComponents: [
    {
      path: '#/channels/channel1/smartylighting/event/{streetlightId}/lighting/measured/message/payload/properties/sentAt',
      action: 'ref',
      refTo: '#/components/schemas/sentAt'
    }
  ],
  RemoveComponents: [
    {
      path: '#/components/messages/unusedMessage',
      action: 'remove',
    }
  ],
  MoveToComponents: [
    {
      //move will ref the current path to the moved component as well.
      path: '#/channels/smartylighting/event/{streetlightId}/lighting/measured/parameters/streetlightId',
      action: 'move',
      moveTo: '#/components/parameters/streetlightId'
    },
    {
      path: '#/channels/smartylighting/action/{streetlightId}/turn/on/parameters/streetlightId',
      action: 'ref',
      refTo: '#/components/parameters/streetlightId'
    }
  ]
}
 */
let optimizedDocument = optimizer.getOptimizedDocument({rules: {ReuseComponents: true,RemoveComponents: true,MoveToComponents: true }})
/*
the `optimizedDocument` value will be:

asyncapi: 2.0.0
info:
  title: Streetlights API
  version: '1.0.0'

channels:

  smartylighting/event/{streetlightId}/lighting/measured:
    parameters:
      streetlightId:
        $ref: "#/components/schemas/streetlightId"
    subscribe:
      operationId: receiveLightMeasurement
      traits:
        - bindings:
            kafka:
              clientId: my-app-id
      message:
        name: lightMeasured
        title: Light measured
        contentType: application/json
        traits:
          - headers:
              type: object
              properties:
                my-app-header:
                  type: integer
                  minimum: 0
                  maximum: 100
        payload:
          type: object
          properties:
            lumens:
              type: integer
              minimum: 0
            #full form is used, we can ref it to: #/components/schemas/sentAt
            sentAt:
              $ref: "#/components/schemas/sentAt"

  smartylighting/action/{streetlightId}/turn/on:
    parameters:
      streetlightId:
        $ref: "#/components/schemas/streetlightId"
    publish:
      operationId: turnOn
      traits:
        - bindings:
            kafka:
              clientId: my-app-id
      message:
        name: turnOnOff
        title: Turn on/off
        traits:
          - headers:
              type: object
              properties:
                my-app-header:
                  type: integer
                  minimum: 0
                  maximum: 100
        payload:
          type: object
          properties:
            sentAt:
              $ref: "#/components/schemas/sentAt"

components:
  parameters:
    streetlightId:
      schema:
      type: string
  schemas:
    #this schema is ref-ed in one channel and used full form in another. library should be able to identify and ref the second channel as well.
    sentAt:
      type: string
      format: date-time`
 */
```

#### new Optimizer(document)

`document` is a mandatory object that is parsed with `@asyncapi/parser`:

#### getOptimizedDocument([options]) : string
* `options`: this is an object that contains all of the options that we can pass in `getOptimizedDocument`. right now we have only `rules` here but maybe in the future we can have other options as well.
  * `rules`: using rules object different optimization types can be enabled or disabled.
    * `ReuseComponents` if set to true, optimizer will apply all of *ReuseComponents* changes from report. (default: *true*)
    * `RemoveComponents` if set to true, optimizer will apply all of *RemoveComponents* changes from report. (default: *false*) 
    * `MoveToComponents` if set to true, optimizer will apply all of *MoveToComponents* changes from report. (default: *true*)

### PR DESCRIPTION
**Pull request by KhudaDad414**
**Description**

Following the discussion in #4, I've modified the API of optimizer library.

**Related issue(s)**
Closes #4

### CODE REVIEW

**Review by derberg — APPROVED** [review]
LGTM
@jazzyarchitects @magicmatatjahu wanna have a look?

File: README.md
Comment (magicmatatjahu) [inline README.md:217]:
```suggestion
`document` is a mandatory object which is the parsed AsyncAPI specification from `@asyncapi/parser`:
```

File: README.md
Comment (magicmatatjahu) [inline README.md:227]: In the https://github.com/asyncapi/optimizer/issues/4 issue you have set true as default value for `reuseComponents` and `moveToComponents`. Why you change it? I think that everything (by default) should be set to true.

File: README.md
Comment (magicmatatjahu) [inline README.md:9]: You use the `Report` type below so you should also import it :)

```suggestion
import { Optimizer } from '@asyncapi/optimizer';
import type { Report } from '@asyncapi/optimizer';
```

**Review by magicmatatjahu — CHANGES_REQUESTED** [review]
Only some comments, overall awesome!

File: README.md
Comment (KhudaDad414) [reply README.md:227]: Sorry about that, I meant to make them true.😆

**Review by magicmatatjahu — APPROVED** [review]
LGTM! Please wait for review from Jabin :)

**Review by jazzyarchitects — CHANGES_REQUESTED** [review]

**Comment by KhudaDad414** [comment]
@jazzyarchitects thanks for the review. :) All of your suggested changes have been applied.

**Comment by jazzyarchitects** [comment]
@KhudaDad414 LGTM

**Review by jazzyarchitects — APPROVED** [review]

**Review by derberg — APPROVED** [review]

**Pull request merged**
